### PR TITLE
fix(native): add prototype method extraction to Rust engine (#1327)

### LINKED ISSUE #1327: fix(native): add prototype method extraction to Rust engine (parity with WASM #1317)
## Context

PR implementing issue #1317 added prototype-based method extraction to the WASM JS extractor:
- `Foo.prototype.bar = function(){}` → emits `Foo.bar` as a method definition
- `Foo.prototype.bar = identifier` → seeds `typeMap['Foo.bar'] = { type: identifier }`
- `Foo.prototype = { bar: fn }` → emits method definitions per property

The call resolver was also extended with:
- Prototype alias lookup: after symbol-DB miss for `TypeName.methodName`, checks `typeMap['TypeName.methodName']`
- Inline `new Foo()` receiver extraction: regex `^\(?\s*new\s+([A-Z_$]...)` extracts the class name when receiver is `(new Foo)` or `(new Foo())`

## What's missing in native

The Rust extractor (`crates/codegraph-core/`) does not implement any of the above. Until it does, WASM and native will produce different results for files containing prototype-based OOP patterns.

## What needs to be done

1. In the native Rust extractor, detect the three prototype assignment patterns and:
   - Emit method definitions for function expressions
   - Emit typeMap entries for identifier aliases
2. In the native call resolver (or its WASM post-processing path), add the prototype alias lookup and inline new-expression receiver handling that were added to `call-resolver.ts`

## References
- Closes when native parity is achieved for issue #1317
- WASM implementation: PR implementing #1317
- Architecture: `crates/codegraph-core/src/extractors/` + native build pipeline

### PR DESCRIPTION
**Pull request by carlos-alm**
## Summary

- Port the pre-ES6 prototype OOP extraction (from PR #1331 / issue #1317) to the native Rust engine to achieve full parity
- Add `match_js_prototype_methods` to `crates/codegraph-core/src/extractors/javascript.rs` — handles all three patterns (`Foo.prototype.bar = fn`, `Foo.prototype.bar = identifier`, `Foo.prototype = { bar: fn }`)
- Add inline `new`-expression receiver extraction and prototype alias lookup to `crates/codegraph-core/src/edge_builder.rs` to match the `call-resolver.ts` resolver path

## Test plan

- [ ] 6 new Rust unit tests in `javascript.rs` (all pass: `cargo test prototype`)
- [ ] Full native test suite — 337/337 pass (`cargo test`)
- [ ] Integration test `tests/integration/prototype-method-resolution.test.ts` — 3/3 pass with locally built native binary
- [ ] Full vitest suite — 2869/2869 pass (`npm test`)

Closes #1327

### CODE REVIEW

File: crates/codegraph-core/src/edge_builder.rs [hunk 1]
```diff
@@ -477,6 +500,31 @@ fn resolve_call_targets<'a>(
...
+/// Extract the constructor name from an inline `new` receiver expression.
+///
+/// Mirrors the regex `/^\(?\s*new\s+([A-Z_$][A-Za-z0-9_$]*)/` used in call-resolver.ts.
+/// Handles `(new Foo)` and `(new Foo('arg'))` receivers that arise when the call site
+/// is `(new Foo).method()` without a named variable binding.
+///
+/// Only extracts PascalCase (uppercase-initial) names to avoid false positives on
+/// lowercase constructor calls (rare but present in legacy code).
+fn extract_inline_new_type(receiver: &str) -> Option<String> {
+    let s = receiver.trim_start_matches('(').trim_start();
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply crates/codegraph-core/src/edge_builder.rs]: Fixed — replaced `trim_start_matches('(')` with `strip_prefix('(').unwrap_or(receiver)` so at most one leading paren is stripped, exactly matching the JS regex `^\(?`. Committed in 0c26030.

File: crates/codegraph-core/src/edge_builder.rs [hunk 2]
```diff
@@ -477,6 +500,31 @@ fn resolve_call_targets<'a>(
...
 }
 
+/// Extract the constructor name from an inline `new` receiver expression.
+///
+/// Mirrors the regex `/^\(?\s*new\s+([A-Z_$][A-Za-z0-9_$]*)/` used in call-resolver.ts.
+/// Handles `(new Foo)` and `(new Foo('arg'))` receivers that arise when the call site
+/// is `(new Foo).method()` without a named variable binding.
+///
+/// Only extracts PascalCase (uppercase-initial) names to avoid false positives on
+/// lowercase constructor calls (rare but present in legacy code).
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply crates/codegraph-core/src/edge_builder.rs]: Fixed — updated the doc comment to accurately describe that names starting with uppercase letter, `_`, or `$` are accepted, matching the JS regex `[A-Z_$]`. Committed in 0c26030.

**Comment by carlos-alm** [comment]
Addressed all review feedback:

**Greptile P2 fixes (both fixed in commit 0c26030):**
- `trim_start_matches('(')` → `strip_prefix('(').unwrap_or(receiver)` to strip at most one paren, matching the JS regex `^\(?`
- Doc comment updated to reflect that `_` and `$` prefixes are also accepted (matching `[A-Z_$]`)

**Out-of-scope findings tracked as issues:**
- Prototype method definitions skip `compute_all_metrics`/`build_function_cfg` → #1340
- `extractPrototypeMethodsWalk` called twice in TS pipeline (duplicate definitions) → #1341

**Comment by carlos-alm** [comment]
@greptileai

**Comment by carlos-alm** [comment]
Follow-up fixes pushed in this batch:

**Greptile string-key stripping bug (commit 41d7811)**
Replaced `.replace([\'\'', '"'], "")` with `strip_prefix`/`strip_suffix` to strip exactly one matching surrounding quote pair. `trim_matches` (suggested in review) would also strip embedded quotes; `strip_prefix`+`strip_suffix` is the correct fix.

**Duplicate `extractPrototypeMethodsWalk` calls (commit a004e25)**
Removed the duplicate calls from both `extractSymbolsQuery` and `extractSymbolsWalk`. Both functions had the call appended a second time at the bottom after it already appeared earlier. This was causing prototype methods to be inserted twice into the definitions array — a WASM parity divergence vs native. Closes #1341.

**Benchmark regression fix (commit 5baf98c)**
CI reported `Build ms/file: 3.6 → 18.8 (+422%)`. Root cause: `runPostNativePrototypeMethods` was still WASM-reparsing all files containing `.prototype.` on every native build, even though the Rust engine now extracts those patterns natively. Removed `.prototype.` from the post-pass file filter — the function-as-object-property pattern (`fn.method = function(){}`) still needs WASM coverage and is retained. Local validation confirms 2.6ms/file full build (within baseline).

**Base branch merge (commit 2fa6d7e + 8dcc760)**
Merged `feat/prototype-resolver-1317` to pick up Phase 8.3f object-rest-param and object-property binding extraction. Fixed format violations inherited from the merge.

**Comment by carlos-alm** [comment]
Fixed the pre-filter regex in native-orchestrator.ts (commit 4e579b0):

The regex `/\b\w+\.\w+\s*=\s*function/` was matching `prototype.bar = function` inside `Foo.prototype.bar = function(){}` because it matched the right-hand portion as a substring. This meant every prototype file was still being queued for WASM re-processing, which Rust now handles natively — creating a risk of duplicate method nodes.

Fixed by adding a negative lookahead: `/\b(?!prototype\.)\w+\.\w+\s*=\s*function/`. This matches only function-as-object-property patterns like `fn.method = function` while excluding prototype assignments, as the comment already described as the intent.

**Comment by carlos-alm** [comment]
@greptileai

**Comment by carlos-alm** [comment]
Addressed Claude's review concerns:

- **Missing `extract_inline_new_type` unit tests (commit 074ced4)**: Added 8 unit tests covering all edge cases flagged in the review: `(new Foo)`, `(new Foo('arg'))`, no-parens form, `_` and `$` prefixes, lowercase rejection, plain identifier, and the `newFoo`-not-a-keyword case.

- **Pre-filter regex still matching prototype files (commit 4e579b0)**: Fixed `/\b\w+\.\w+\s*=\s*function/` → `/\b(?!prototype\.)\w+\.\w+\s*=\s*function/` with a negative lookahead, so prototype-assignment files are excluded from the WASM post-pass as intended.

- **Missing complexity metrics for prototype method definitions**: Tracked as issue #1340. Out of scope for this PR — this PR targets call-edge parity, and the TS reference also emits `null` complexity for these symbols (the gap exists in both engines equally).

- **No dedicated unit test for the prototype-alias fallback path in `edge_builder.rs`**: The integration test `prototype-method-resolution.test.ts` covers this path end-to-end. Adding a focused Rust unit test for the fallback would require mocking the DB and SQLite context — I'll track that as a follow-up in issue #1340 alongside the complexity metrics work.

**Comment by carlos-alm** [comment]
@claude

**Comment by carlos-alm** [comment]
Fixed CI failures (commit fceb2bf):

The JS benchmark was failing precision (84.4% < 100%) and recall (81.8% < 90%) due to a stale `expected-edges.json` manifest. Root cause: commit `4ed709e` applied biome auto-fixes that renamed fixture functions `defProp/defProps/create` → `_defProp/_defProps/_create` (unused-variable convention), but the manifest still referenced the old names — producing 5 FP and 5 FN.

Fixes applied:
1. Updated `expected-edges.json` to reference `_defProp`, `_defProps`, `_create` (matching the current fixture).
2. Fixed `inheritance.js` DoubleCounter fixture to use `super.count()` instead of `Counter.count()` so it exercises the class-inheritance resolution path the manifest documents.
3. Added `biome.json` override to disable lint on `tests/benchmarks/resolution/fixtures/**` — benchmark fixtures are hand-written sample code that must use specific patterns (unused top-level functions, super calls) to test resolution. This prevents future recurrence.

Local test results: precision=100%, recall=93.9% (all JS benchmark tests pass).

**Comment by carlos-alm** [comment]
@greptileai
